bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        probabilistic: bool = True,
+        dampening: float = 1.0,
+    ):
+        """Initialize the DrGhmcDiag sampler.
+
+        Args:
+            model: model with log density and gradient
+            stepsize: stepsize in each leapfrog step. Defaults to None.
+            steps: number of leapfrog steps. Defaults to None.
+            metric_diag: diagonal of the metric. Defaults to None.
+            init: initialize Markov chain. Defaults to None.
+            seed: seed for numpy rng. Defaults to None.
+            num_proposals: number of delayed rejection proposals. Defaults to 3.
+            probabilistic: whether to use probabilistic delayed rejection. Defaults to True.
+            dampening: GHMC momentum dampening factor in [0, 1] inclusive. Defaults to 1.0.
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        self._num_proposals = num_proposals
+        self._stepsize_list = self._init_stepsize(stepsize)
+        self._steps_list = self._init_steps(steps)
+        self._metric = metric_diag or np.ones(self._dim)
+        self._rng = np.random.default_rng(seed)
+        self._theta = (
+            init
+            if (init is not None and init.shape != (0,))
+            else self._rng.normal(size=self._dim)
+        )
+        self._rho = self._rng.normal(size=self._dim)
+        self._probabilistic = probabilistic
+        if not 0 <= dampening <= 1:
+            raise ValueError(f"dampening must be within [0, 1] inclusive")
+        self._dampening = dampening
+
+    def _init_stepsize(
+        self,
+        stepsize: Union[int | float, list[int | float], Callable[[int], int | float]],
+    ) -> list[float]:
+        """List with leapfrog stepsizes used to generate each proposed sample.
+
+        Each proposed sample (theta_prop, rho_prop) is generated by taking multiple
+        leapfrog steps with a a fixed stepsize. Each distinct proposal, however, may
+        have a different stepsize. This function initializes the stepsize for
+        self._num_proposal such proposals.
+
+        This function accepts, among other valid inputs, a callable stepsize function.
+        This callable should accept the (integer) proposal number and return the
+        leapfrog stepsize for that proposal.
+
+        Args:
+            stepsize: stepsize in each leapfrog step
+
+        Args:
+            stepsize: done
+
+        Raises:
+            ValueError: scalar with non-positive stepsize
+            ValueError: list with incorrect number of stepsizes
+            ValueError: list with non-positive stepsize
+            TypeError: list with non-numeric (int, float) stepsizes
+            ValueError: callable returns non-positive stepsize
+            TypeError: callable returns non-numeric (int, float) stepsize
+            TypeError: invalid type for stepsize
+
+        Returns:
+            list[float]: leapfrog stepsizes for self._num_proposal proposals
```

Comment:
As discussed elsewhere, the function's return type should be annotated on the function directly rather than in the docstring.